backend/telemetry_core/src/find_location.rs

```diff
@@ -110,98 +101,25 @@ impl Locator {
             return cached_loc;
         }
 
-        // Look it up via ipapi.co:
-        let mut location = self.iplocate_ipapi_co(ip).await;
-
-        // If that fails, try looking it up via ipinfo.co instead:
-        if let Err(e) = &location {
-            log::warn!(
-                "Couldn't obtain location information for {} from ipapi.co: {}",
-                ip,
-                e
-            );
-            location = self.iplocate_ipinfo_io(ip).await
-        }
-
-        // If both fail, we've logged the errors and we'll return None.
-        if let Err(e) = &location {
-            log::warn!(
-                "Couldn't obtain location information for {} from ipinfo.co: {}",
-                ip,
-                e
-            );
-        }
-
-        // If we successfully obtained a location, cache it
-        if let Ok(location) = &location {
-            self.cache.write().insert(ip, location.clone());
-        }
-
-        // Discard the error; we've logged information above.
-        location.ok()
-    }
-
-    async fn iplocate_ipapi_co(&self, ip: Ipv4Addr) -> Result<Arc<NodeLocation>, anyhow::Error> {
-        let location = self.query(&format!("https://ipapi.co/{}/json", ip)).await?;
-
-        Ok(Arc::new(location))
-    }
-
-    async fn iplocate_ipinfo_io(&self, ip: Ipv4Addr) -> Result<Arc<NodeLocation>, anyhow::Error> {
-        let location = self
-            .query::<IPApiLocate>(&format!("https://ipinfo.io/{}/json", ip))
-            .await?
-            .into_node_location()
-            .with_context(|| "Could not convert response into node location")?;
-
-        Ok(Arc::new(location))
-    }
-
-    async fn query<T>(&self, url: &str) -> Result<T, anyhow::Error>
-    where
-        for<'de> T: Deserialize<'de>,
-    {
-        let res = self
-            .client
-            .get(url)
-            .send()
-            .await?
-            .bytes()
-            .await
-            .with_context(|| "Failed to obtain response body")?;
-
-        serde_json::from_slice(&res)
-            .with_context(|| format!{"Failed to decode '{}'", std::str::from_utf8(&res).unwrap_or("INVALID_UTF8")})
-    }
-}
-
-/// This is the format returned from ipinfo.co, so we do
-/// a little conversion to get it into the shape we want.
-#[derive(Deserialize, Debug, Clone)]
-struct IPApiLocate {
-    city: Box<str>,
-    loc: Box<str>,
-}
-
-impl IPApiLocate {
-    fn into_node_location(self) -> Option<NodeLocation> {
-        let IPApiLocate { city, loc } = self;
-
-        let mut loc = loc.split(',').map(|n| n.parse());
-
-        let latitude = loc.next()?.ok()?;
-        let longitude = loc.next()?.ok()?;
-
-        // Guarantee that the iterator has been exhausted
-        if loc.next().is_some() {
-            return None;
-        }
-
-        Some(NodeLocation {
+        let City { city, location, .. } = self.city.lookup(ip.into()).ok()?;
```

Comment:
This needs to happen in a separate blocking task, it is too expensive to resolve within async context. This causes issues when deployed at scale.